src/mon/PGMap.cc

```diff
@@ -2363,6 +2363,26 @@ void PGMap::dump_pool_stats_and_io_rate(int64_t poolid, const OSDMap &osd_map,
   }
 }
 
+// Get crush parentage for an osd (skip root)
+set<std::string> PGMap::osd_parentage(const OSDMap& osdmap, int id) const
+{
+  set<std::string> reporters_by_subtree;
```

Comment:
is it possible that an OSD have the same parent for multiple times?

Reply:
Yes, but in real life these errors should be rare and I don't think it is worth caching the results for efficiency.